cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+			found = true
+		} else {
+			newMigs = append(newMigs, mig)
+		}
+	}
+	gc.migs = newMigs
+	return found
+}
+
+// GetMigs returns a copy of migs list.
+func (gc *GceCache) GetMigs() []*MigInformation {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	migs := make([]*MigInformation, 0, len(gc.migs))
+	for _, mig := range gc.migs {
+		migs = append(migs, &MigInformation{
+			Basename: mig.Basename,
+			Config:   mig.Config,
+		})
+	}
+	return migs
+}
+
+func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == ref {
+			mig.Basename = basename
+		}
+	}
+}
+
+// Methods locking on cacheMutex.
+
+// GetMigForInstance returns MigConfig of the given Instance from cache.
+// Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
+// TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
+// matching prefix, but instance doesn't belong to it.
+func (gc *GceCache) GetMigForInstance(instance *GceRef) (Mig, error) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	if mig, found := gc.migCache[*instance]; found {
+		return mig, nil
+	}
+
+	for _, mig := range gc.GetMigs() {
+		if mig.Config.GceRef().Project == instance.Project &&
+			mig.Config.GceRef().Zone == instance.Zone &&
+			strings.HasPrefix(instance.Name, mig.Basename) {
+			if err := gc.regenerateCache(); err != nil {
+				return nil, fmt.Errorf("Error while looking for MIG for instance %+v, error: %v", *instance, err)
+			}
+			if mig, found := gc.migCache[*instance]; found {
+				return mig, nil
+			}
+			return nil, fmt.Errorf("Instance %+v does not belong to any configured MIG", *instance)
+		}
+	}
+	// Instance doesn't belong to any configured mig.
+	return nil, nil
+}
+
+// RegenerateCacheWithLock triggers cache regeneration under lock.
+func (gc *GceCache) RegenerateCacheWithLock() error {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.regenerateCache()
+}
+
+// internal method - should only be called after locking on cacheMutex.
+func (gc *GceCache) regenerateCache() error {
+	newMigCache := make(map[GceRef]Mig)
+
+	for _, migInfo := range gc.GetMigs() {
+		mig := migInfo.Config
+		glog.V(4).Infof("Regenerating MIG information for %s", mig.GceRef().String())
+
+		basename, err := gc.GceService.FetchMigBasename(mig.GceRef())
+		if err != nil {
+			return err
+		}
+		gc.updateMigBasename(mig.GceRef(), basename)
+
+		instances, err := gc.GceService.FetchMigInstances(mig.GceRef())
+		if err != nil {
+			glog.V(4).Infof("Failed MIG info request for %s: %v", mig.GceRef().String(), err)
+			return err
+		}
+		for _, ref := range instances {
+			newMigCache[ref] = mig
+		}
+	}
+
+	gc.migCache = newMigCache
+	return nil
+}
+
+// SetResourceLimiter sets resource limiter.
+func (gc *GceCache) SetResourceLimiter(resourceLimiter *cloudprovider.ResourceLimiter) {
```

Comment:
This has really nothing to do with this cache, except we happen to have some mutex in cache that we could use to protect it.

Reply:
We used this same lock to protect resource limits before. Also, I'd argue that if the intended behavior is to fetch limits from the API, but we don't want to do this too often, so we keep them and refresh only now and then, it technically makes it a cached value.